pkg/services/baremetal/host/host.go

```diff
@@ -951,14 +953,54 @@ func (s *Service) actionImageInstalling() actionResult {
 
 	s.scope.HetznerBareMetalHost.Spec.Status.SSHStatus.OSKey = &sshKey
 
+	// If there is a Linux OS on an other disk, then the reboot after the provisioning
+	// will likely fail, because the machine boots into the other operating system.
+	// We want detect that early, and not start the provisioning process.
+	out := sshClient.DetectLinuxOnOtherDisk(s.scope.HetznerBareMetalHost.Spec.RootDeviceHints.ListOfWWN())
+	if out.Err != nil {
+		var exitErr *ssh.ExitError
+		if errors.As(out.Err, &exitErr) && exitErr.ExitStatus() > 0 {
+			// The script detected Linux on an other disk. This is a permanent error.
+			msg := fmt.Sprintf("DetectLinuxOnOtherDisk failed (permanent error): %s. StdErr: %s (%s)",
+				out.StdOut, out.StdErr, out.Err.Error())
+			conditions.MarkFalse(
+				s.scope.HetznerBareMetalHost,
+				infrav1.ProvisionSucceededCondition,
+				infrav1.LinuxOnOtherDiskFoundReason,
+				clusterv1.ConditionSeverityError,
+				msg,
+			)
+			record.Warn(s.scope.HetznerBareMetalHost, infrav1.LinuxOnOtherDiskFoundReason, msg)
+			s.scope.HetznerBareMetalHost.SetError(infrav1.PermanentError, msg)
+			return actionStop{}
+		}
+
+		// Some other error like connection timeout. Retry again later.
+		// This often during provisioning.
+		msg := fmt.Sprintf("DetectLinuxOnOtherDisk failed (will retry): %s. StdErr: %s (%s)",
+			out.StdOut, out.StdErr, out.Err.Error())
+		conditions.MarkFalse(
+			s.scope.HetznerBareMetalHost,
+			infrav1.ProvisionSucceededCondition,
+			infrav1.SSHToRescueSystemFailedReason,
+			clusterv1.ConditionSeverityInfo,
+			msg,
+		)
+		record.Event(s.scope.HetznerBareMetalHost, infrav1.SSHToRescueSystemFailedReason, msg)
+		return actionContinue{
+			delay: 10 * time.Second,
+		}
+	}
+	conditions.MarkTrue(s.scope.HetznerBareMetalHost, infrav1.ProvisionSucceededCondition)
```

Comment:
did the provisioning really succeed here?

Reply:
@janiskemper 

no, but if you don't set it to OK here, then the error might be visibly longer than it is there. I removed that line. I think creating a custom condition is not needed, and if the err is visibly longer than it exists is not a big issue.